### DISCUSSION TITLE
Creating PDFs

### DISCUSSION THREAD

**Discussion by JamesWanjohi** [tethysplatform/tethys#670] [Q&A]
Hello.

I would like help on a workflow to creating PDFs from data overlaid from Google Earth Engine (GEE) on a Tethys platform map canvas.

Regards,
James

**Comment by swainn** [comment]
@JamesWanjohi thank you for your question. Generating PDFs is not as simple as you think it would be, but here are some pointers that I've found helpful. There are two approaches that I've found for doing this:

1. Create a print-able version of your page and then use the "Save as PDF" functionality that many web browsers support in the Print dialog for a web page. This is the simplest of the two methods b/c it leverages capabilities common to web browsers. Here's the high-level overview of how to do this:

    - Add a button to your page that calls a JavaScript function when you click on it. I usually title the button "Print" or "Print Friendly".
    - Using JavaScript, simplify the page contents by hiding many of the elements such as the Header, footer, navigation. Also, resize elements so they will fit on the width of a standard page. You should only have the content you want in the PDF when you are done (i.e. the map)
    - Open the print dialog using `window.print();`. The user will see the native print dialog open on their web browser. Depending on the browser, one of the options is usually to save as a PDF.
    - When printing is done, restore the page to what it was before. A simple way to do this is to reload the page.

2. Create a PDF document on-the-fly using a JavaScript library like `jsPDF`. There are many of these, just Google "JavaScript PDF Libraries" and find one that suits your needs. Here's a high-level overview of this approach:

    - Add a button to your page that calls a javascript function when you click on it with a title like "Download" or "Save as PDF".
    - In the JavaScript function, use the PDF generation library to create a new PDF document and the content that you wish to it.
    - The trickiest part of this will be getting the content of the map. Generally, you want to find a way to convert the rendered map into an image. There are different approaches to this, depending on the mapping library you are using.
        - Leaflet Maps: there are several plugins you can add to support exporting the map as an image or even creating a print friendly version of it. See: https://leafletjs.com/plugins.html#printexport
        - OpenLayers Maps I found this example to convert a map to an image: https://openlayers.org/en/latest/examples/export-map.html
    - Once you have the map as an image, you can easily add the image to the new PDF document.

**Reply by JamesWanjohi** [reply]
Thank you so much for this information.

I will have a go and revert.

Kind regards,
James
